File: maxSubArray/maxSubArray.py
#!/usr/bin/env python


class Solution:
    # @param A, a list of integers
    # @return an integer

    # O(n)
    def maxSubArray(self, A):
        if len(A) == 0:
            return 0

        max_ending_here = max_so_far = A[0]
        for x in A[1:]:
            max_ending_here = max(x, max_ending_here + x)
            max_so_far = max(max_so_far, max_ending_here)
        return max_so_far


    # divide and conquer
#    def maxSubArray(self, A):
        #if len(A) == 1:
            #return A[0]

        #mid = len(A)/2
        #left = A[:mid]
        #right = A[mid:]

        #max_left = self.maxSubArray(left)
        #max_right = self.maxSubArray(right)
        #max_border = self.maxSum(reversed(left)) + self.maxSum(right)

        #return max(max_left, max_right, max_border)

    #def maxSum(self, arrayB):
        #"""docstring for maxSum"""
        #n_max = float('-inf')
        #n_sum = 0
        #for item in arrayB:
            #n_sum += item
            #if n_sum > n_max:
                #n_max = n_sum
        #return n_max


def test():
    """docstring for test"""

    iList = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
    sol = Solution()
    print sol.maxSubArray(iList)


if __name__ == '__main__':
    test()

File: maxSubArray/maxSubArray.cpp
#include <iostream>
#include <algorithm>
#include <climits>

class Solution {
public:
    // O(n)
    //int maxSubArray(int A[], int n) {
        //if (n == 0) {
            //return 0;
        //}

        //int max_ending_here, max_so_far;
        //max_ending_here = max_so_far = A[0];

        //for (int i = 1; i < n; i++) {
            //max_ending_here = std::max(A[i], max_ending_here + A[i]);
            //max_so_far = std::max(max_so_far, max_ending_here);
        //}
        //return max_so_far;
    //}

    // divide and conquer
    int maxSubArray(int A[], int n) {
        return maxSubArray(A, 0, n-1);
    }

    int maxSubArray(int A[], int left, int right) {
        if (left > right) {
            return INT_MIN;
        }
        if(left == right) {
            return A[left];
        }
        int mid = left + (right - left)/2;
        int max_left = maxSubArray(A, left, mid);
        int max_right = maxSubArray(A, mid+1, right);

        int max_left_bd = INT_MIN;
        int sum_left_bd = 0;
        for(int i= mid; i >=left; i--) {
            sum_left_bd += A[i];
            if (sum_left_bd > max_left_bd) {
                max_left_bd = sum_left_bd;
            }
        }

        int max_right_bd = INT_MIN;
        int sum_right_bd = 0;
        for (int j = mid + 1; j <=right; j++) {
            sum_right_bd += A[j];
            if (sum_right_bd > max_right_bd) {
                max_right_bd = sum_right_bd;
            }
        }
        int max_bd = max_left_bd + max_right_bd;
        return (max_left > max_right)? ((max_left > max_bd)? max_left: max_bd):((max_right > max_bd)? max_right: max_bd);
    }
};


int main(int argc, const char *argv[])
{
    Solution *sl = new Solution();
    int A[] = {-2,1,-3,4,-1,2,1,-5,4};
    int A_len = 9;

    std::cout << sl->maxSubArray(A, A_len) << std::endl;

    return 0;
}
